# Contents
1. Imports
2. Checks
3. Wrangling
4. Cleaning
5. Exports

# 1. Imports

In [2]:
#Libraries
import pandas as pd
import numpy as np
import os

In [3]:
#Path
path = r'/Users/davidgriesel/Documents/GitHub/202409_OGS'

In [4]:
#Dataset
df_orders_products = pd.read_csv(os.path.join(path, '02 - Data', 'Original Data', 'orders_products.csv'))

# 2. Checks

In [5]:
#Dimensions
df_orders_products.shape

(32434489, 4)

In [6]:
#Preview
df_orders_products.head()

order_id  product_id  add_to_cart_order  reordered
0         2       33120                  1          1
1         2       28985                  2          1
2         2        9327                  3          0
3         2       45918                  4          1
4         2       30035                  5          0

In [7]:
#Summary stats
df_orders_products.describe()

order_id    product_id  add_to_cart_order     reordered
count  3.243449e+07  3.243449e+07       3.243449e+07  3.243449e+07
mean   1.710749e+06  2.557634e+04       8.351076e+00  5.896975e-01
std    9.873007e+05  1.409669e+04       7.126671e+00  4.918886e-01
min    2.000000e+00  1.000000e+00       1.000000e+00  0.000000e+00
25%    8.559430e+05  1.353000e+04       3.000000e+00  0.000000e+00
50%    1.711048e+06  2.525600e+04       6.000000e+00  1.000000e+00
75%    2.565514e+06  3.793500e+04       1.100000e+01  1.000000e+00
max    3.421083e+06  4.968800e+04       1.450000e+02  1.000000e+00

##### Observations:
- Dataset contains 32,434,489 records
- Possible missing records in order_id with min = 2

# 3. Wrangling

## 3.1. Drop Columns

### 3.1.1. Identify Redundant Columns

In [8]:
#Check frequency distribution of variable
df_orders_products['order_id'].value_counts(dropna = False)

order_id
1564244    145
790903     137
61355      127
2970392    121
2069920    116
          ... 
118997       1
118995       1
1119343      1
870412       1
2530309      1
Name: count, Length: 3214874, dtype: int64

##### Observations:
- 3,214,874 unique values ranging between 2 to 3,421,083
- With 32,434,489 records the dataset likely has missing records
- Retain column

In [9]:
#Check frequency distribution of variable
df_orders_products['product_id'].value_counts(dropna = False)

product_id
24852    472565
13176    379450
21137    264683
21903    241921
47209    213584
          ...  
14756         1
20264         1
31254         1
13397         1
23624         1
Name: count, Length: 49677, dtype: int64

##### Observations:
- 49,677 values ranging between 1 and 49,688
- Indicates missing records and/or that certain products have never been ordered
- Retain column

In [10]:
#Check frequency distribution of variable
df_orders_products['add_to_cart_order'].value_counts(dropna = False)

add_to_cart_order
1      3214874
2      3058126
3      2871133
4      2664106
5      2442025
        ...   
141          1
142          1
143          1
144          1
145          1
Name: count, Length: 145, dtype: int64

##### Observations:
- 145 values ranging between 1 and 145
- Customers ordered up to 145 items at a time
- Retain column

In [11]:
#Check frequency distribution of variable
df_orders_products['reordered'].value_counts(dropna = False)

reordered
1    19126536
0    13307953
Name: count, dtype: int64

##### Observations:
- 2 values with 1 indicating customer have ordered product before, and 0 indicating first time order 
- Retain column

### 3.1.2. Address Redundant Columns

##### Observations:
- No redundant columns identified

## 3.2. Rename Columns

### 3.2.1. Identify Unclear Descriptions

In [12]:
#Check Column Descriptions
df_orders_products.columns

Index(['order_id', 'product_id', 'add_to_cart_order', 'reordered'], dtype='object')

##### Observations:
- Variables are described clearly

### 3.2.2. Address Unclear Descriptions

##### Observations:
- No unclear descriptions noted

## 3.3. Data Types

### 3.3.1. Identify Inconsistent Data Types

In [13]:
#Check data types
df_orders_products.dtypes

order_id             int64
product_id           int64
add_to_cart_order    int64
reordered            int64
dtype: object

##### Observations:
- Data types are consistent with variable contents

### 3.3.2. Address Inconsistent Data Types

##### Observations:
- No inconsistencies noted

# 4. Cleaning

## 4.1. Accuracy

### 4.1.1. Identify Inaccurate Values

In [14]:
#Review descriptive statistics
df_orders_products.describe()

order_id    product_id  add_to_cart_order     reordered
count  3.243449e+07  3.243449e+07       3.243449e+07  3.243449e+07
mean   1.710749e+06  2.557634e+04       8.351076e+00  5.896975e-01
std    9.873007e+05  1.409669e+04       7.126671e+00  4.918886e-01
min    2.000000e+00  1.000000e+00       1.000000e+00  0.000000e+00
25%    8.559430e+05  1.353000e+04       3.000000e+00  0.000000e+00
50%    1.711048e+06  2.525600e+04       6.000000e+00  1.000000e+00
75%    2.565514e+06  3.793500e+04       1.100000e+01  1.000000e+00
max    3.421083e+06  4.968800e+04       1.450000e+02  1.000000e+00

In [15]:
#Number of unique values per variable
df_orders_products.nunique()

order_id             3214874
product_id             49677
add_to_cart_order        145
reordered                  2
dtype: int64

##### Observations:
- The dataset cotains 32,434,489 records
- With 3,214,874 unique user_id's ranging between 2 to 3,421,083 the dataset likely contains duplicate and/or missing records
- With 49,677 unique product_id ranging between 1 and 49,688 it either confirms missing records, or that certain products have never been ordered
- Customers ordered between 1 and 145 products at a time
- Dataset indicates of products have been ordered before (1) or not (0).

### 4.1.2. Address Inaccurate Values

##### Observations
- No inaccuracies noted

## 4.2. Missing Values

### 4.2.1. Identify missing values

In [16]:
#Number of missing values per variable
df_orders_products.isnull().sum()

order_id             0
product_id           0
add_to_cart_order    0
reordered            0
dtype: int64

##### Observations:
- None of the variables had NaN values

### 4.2.2. Address Missing Values

##### Observations:
- No missing values were identified

## 4.3. Mixed Type Variables

### 4.3.1. Find Mixed Type Variables

In [17]:
#Finding mixed type data
for col in df_orders_products.columns.tolist():
    weird = (df_orders_products[[col]].map(type) != df_orders_products[[col]].iloc[0].apply(type)).any(axis=1)
    if len(df_orders_products[weird]) > 0:
        print (col)

##### Observations:
- No variables returned

### 4.3.2. Address Mixed Type Variables

##### Observations:
- No variables identified with with mixed type data

## 4.4. Duplicates

### 4.4.1. Find Duplicates

In [18]:
#Identify duplicates, create and view subset with results
df_orders_products_duplicates = df_orders_products[df_orders_products.duplicated()]
df_orders_products_duplicates

Empty DataFrame
Columns: [order_id, product_id, add_to_cart_order, reordered]
Index: []

In [19]:
#Confirm dimensions
df_orders_products_duplicates.shape

(0, 4)

##### Observations:
- No records returned

### 4.4.2. Address Duplicates

##### Observations
- No duplicate records identified

# 5. Exports

In [20]:
#Confirm dimensions
df_orders_products.shape

(32434489, 4)

In [21]:
#Export cleaned dataset
df_orders_products.to_pickle(os.path.join(path, '02 - Data', 'Prepared Data', '04_cleaned_orders_products.pkl'))